Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF

cdumez opened the pull request:
#### 255643808bacd89e934b7100c04b6ef875005a02
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF
<a href="https://bugs.webkit.org/show_bug.cgi?id=283636">https://bugs.webkit.org/show_bug.cgi?id=283636</a>

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringImpl):
(WTF::writeIntegerToBufferImpl):
* Source/WTF/wtf/text/MakeString.h:
* Source/WTF/wtf/text/StringBuilderInternals.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::span):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::stringLength): Deleted.

Canonical link: <a href="https://commits.webkit.org/287040@main">https://commits.webkit.org/287040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ccb7209661852e8b1fdce6c743e9a888f7ca43f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82696 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29302 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5374 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61116 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19032 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81114 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/67486 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41428 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24583 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27647 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71190 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24934 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84058 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77282 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5413 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3651 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5569 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66969 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68585 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12586 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10776 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99594 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12083 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5361 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21758 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5351 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8783 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -65,6 +65,23 @@ inline std::span<const char> span(const char* string)
+}
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+inline std::span<const UChar> span(const UChar* string)
```
I guess in the long run to make this safe we’ll need a type to distinguish “null terminated pointer” from other kinds of pointer.

ddkilzer replied:
`ASCIILiteral.h` has a similar [issue](https://github.com/WebKit/WebKit/commit/00e4ae32b87a36717be7ed03a9f62cf6bdc456ce#diff-44914e2a33d00f6d7301f76002cae918316e3acd527ac9ee2bac4e27361806bbR52) as I recall.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF
https://bugs.webkit.org/show_bug.cgi?id=283636

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringImpl):
(WTF::writeIntegerToBufferImpl):
* Source/WTF/wtf/text/MakeString.h:
* Source/WTF/wtf/text/StringBuilderInternals.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::span):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::stringLength): Deleted.

Canonical link: https://commits.webkit.org/287040@main
